### PR TITLE
[RFC] Better way too loop over grid cells

### PR DESCRIPTION
**Pull request by fredrikekre**
First take on a better cell iterator interface.

This only provide a way to loop over the grid, without coupling with `CellValues`.

(keeping it in a separate file until ready to be put in `grid.jl` I guess)

### CODE REVIEW

File: src/cell_iterator.jl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+end
+
+function Base.start{dim, N, T}(grid::Grid{dim, N, T})
+    # N is number of nodes per cell in the grid
+    nodes = zeros(Int, N)
+    coords = zeros(Vec{dim, T}, N)
+    return CellIterator(Ref(0), nodes, coords)
+end
+
+function Base.next{dim, N, T}(grid::Grid{dim, N, T}, ci::CellIterator{dim, T})
```
Comment (fredrikekre) [inline src/cell_iterator.jl]: for some reason `next` seem to allocate memory...

**Comment by fredrikekre** [comment]
The alternative approach would be to create a new `CellIterator` in each `next`-call, which would allow state to be just an integer. But that should be less efficient, right?

**Comment by fredrikekre** [comment]
Why is it so much slower to return the same object twice?

returning just one:

```jl
julia> @benchmark next($grid, $ci)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     13.282 ns (0.00% GC)
  median time:      13.442 ns (0.00% GC)
  mean time:        14.544 ns (0.00% GC)
  maximum time:     84.151 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     998
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

returning two:

```jl
julia> @benchmark next($grid, $ci)
BenchmarkTools.Trial: 
  memory estimate:  32.00 bytes
  allocs estimate:  1
  --------------
  minimum time:     19.894 ns (0.00% GC)
  median time:      20.283 ns (0.00% GC)
  mean time:        23.095 ns (5.91% GC)
  maximum time:     1.246 μs (96.98% GC)
  --------------
  samples:          10000
  evals/sample:     996
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by fredrikekre** [comment]
Before latest commit:

```jl
julia> @benchmark next($grid, $ci)
BenchmarkTools.Trial: 
  memory estimate:  2.03 kb
  allocs estimate:  41
  --------------
  minimum time:     1.683 μs (0.00% GC)
  median time:      1.762 μs (0.00% GC)
  mean time:        2.031 μs (5.89% GC)
  maximum time:     147.508 μs (97.36% GC)
  --------------
  samples:          10000
  evals/sample:     10
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by fredrikekre** [comment]
```jl
julia> foo(x) = x;

julia> bar(x) = (x, x);

julia> x = rand(4);

julia> @code_native foo(x)
	.text
Filename: REPL[1]
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 1
	movq	%rdi, %rax
	popq	%rbp
	retq
	nopl	(%rax)

julia> @code_native bar(x)
	.text
Filename: REPL[2]
	pushq	%rbp
	movq	%rsp, %rbp
	pushq	%r14
	pushq	%rbx
	subq	$32, %rsp
	movq	%rdi, %r14
	movq	%fs:0, %rbx
	addq	$-2672, %rbx            # imm = 0xFFFFFFFFFFFFF590
	vxorps	%xmm0, %xmm0, %xmm0
	vmovups	%xmm0, -32(%rbp)
	movq	$4, -48(%rbp)
	movq	(%rbx), %rax
	movq	%rax, -40(%rbp)
	leaq	-48(%rbp), %rax
	movq	%rax, (%rbx)
Source line: 1
	movabsq	$jl_gc_pool_alloc, %rax
	movl	$1456, %esi             # imm = 0x5B0
	movl	$32, %edx
	movq	%rbx, %rdi
	callq	*%rax
	movabsq	$140325128561296, %rcx  # imm = 0x7F9FFD70E290
	movq	%rcx, -8(%rax)
	movq	%rax, -32(%rbp)
	movq	%r14, -24(%rbp)
	movq	%r14, (%rax)
	movq	%r14, 8(%rax)
	movq	-40(%rbp), %rcx
	movq	%rcx, (%rbx)
	addq	$32, %rsp
	popq	%rbx
	popq	%r14
	popq	%rbp
	retq
	nopw	%cs:(%rax,%rax)

julia> using BenchmarkTools

julia> @benchmark foo($x)
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.815 ns (0.00% GC)
  median time:      1.891 ns (0.00% GC)
  mean time:        1.890 ns (0.00% GC)
  maximum time:     42.746 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark bar($x)
BenchmarkTools.Trial: 
  memory estimate:  32.00 bytes
  allocs estimate:  1
  --------------
  minimum time:     7.399 ns (0.00% GC)
  median time:      10.367 ns (0.00% GC)
  mean time:        11.983 ns (10.84% GC)
  maximum time:     1.050 μs (97.74% GC)
  --------------
  samples:          10000
  evals/sample:     999
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by KristofferC** [comment]
You need to allocate memory for the tuple when it contains objects that are allocated on the heap (like arrays).

**Comment by fredrikekre** [comment]
But its is the same array?

**Comment by fredrikekre** [comment]
eg.
```jl
julia> x = rand(4);

julia> bar(x) = (x,x);

julia> a, b = bar(x)
([0.637593,0.308879,0.653693,0.639757],[0.637593,0.308879,0.653693,0.639757])

julia> a[1] = 1
1

julia> b
4-element Array{Float64,1}:
 1.0     
 0.308879
 0.653693
 0.639757

julia> pointer(a)
Ptr{Float64} @0x00007fe1bad04d00

julia> pointer(b)
Ptr{Float64} @0x00007fe1bad04d00
```

**Comment by fredrikekre** [comment]
Anyhow, now we can do this:

```jl
julia> qr = QuadratureRule{3, RefCube}(2);

julia> ip = Lagrange{3, RefCube, 1}();

julia> cv = CellScalarValues(qr, ip);

julia> grid = generate_grid(Hexahedron, (3,3,3));

julia> for cell in grid
           reinit!(cv, cell)
           # do stuff
       end
```

**Comment by KristofferC** [comment]
Yea but the tuple itself needs to be allocated.

**Pull request merged**